grpc-sys/grpc_wrap.cc

```diff
@@ -844,4 +848,134 @@ grpcwrap_ssl_server_credentials_create(
   return creds;
 }
 
+using grpc::CompletionQueue;
+
+/** Wrapper of `grpc::CompletionQueue`.
+ */
+typedef CompletionQueue grpcwrap_completion_queue;
+
+/** Create a `grpcwrap_completion_queue`, a wrapper of `grpc_completion_queue`.
+ */
+GPR_EXPORT grpcwrap_completion_queue* GPR_CALLTYPE
+grpcwrap_completion_queue_create() {
+  grpc_completion_queue* cq = grpc_completion_queue_create_for_next(NULL);
+  CompletionQueue* cq_wrapper = new CompletionQueue(cq);
+  return cq_wrapper;
+}
+
+/** Returns a raw pointer to the underlying a grpc_completion_queue
+ * instance.
+ */
+GPR_EXPORT grpc_completion_queue* GPR_CALLTYPE
+grpcwrap_completion_queue_cq(grpcwrap_completion_queue* cq_wrapper) {
+  grpc_completion_queue* cq = cq_wrapper->cq();
+  return cq;
+}
+
+/** Request the shutdown of the queue.
+ */
+GPR_EXPORT void GPR_CALLTYPE
+grpcwrap_completion_queue_shutdown(grpcwrap_completion_queue* cq) {
+  cq->Shutdown();
+  return;
+}
+
+/** Destroy a `grpcwrap_completion_queue`.
+ */
+GPR_EXPORT void GPR_CALLTYPE
+grpcwrap_completion_queue_destroy(grpcwrap_completion_queue* cq) {
+  delete cq;
+  return;
+}
+
+using grpc::Alarm;
+
+typedef Alarm grpcwrap_alarm;
+
+/** Create a completion queue alarm instance */
+GPR_EXPORT grpcwrap_alarm* GPR_CALLTYPE grpcwrap_alarm_create(void) {
+  Alarm* alarm = new Alarm();
+  return alarm;
+}
+
+/** Set a completion queue alarm instance associated to a cq.
+ *
+ * Once the alarm notifies (see a grpcwrap_alarm_cancel) or it's destroy (see a
+ * grpcwrap_alarm_destroy), an event with a tag will be added to a cq. If the
+ * alarm notified, the event's success bit will be true, false otherwise. */
+GPR_EXPORT void GPR_CALLTYPE grpcwrap_alarm_set(grpcwrap_alarm* alarm,
+                                                grpcwrap_completion_queue* cq,
+                                                void* tag) {
+  gpr_timespec inf = gpr_inf_future(GPR_CLOCK_REALTIME);
+  alarm->Set(cq, inf, tag);
+  return;
+}
+
+/** Notify a completion queue alarm. Calling this function over an alarm that
+ * has already fired has no effect. */
+GPR_EXPORT void GPR_CALLTYPE grpcwrap_alarm_cancel(grpcwrap_alarm* alarm) {
+  alarm->Cancel();
+  return;
+}
+
+/** Destroy the given completion queue alarm, cancelling it in the process. */
+GPR_EXPORT void GPR_CALLTYPE grpcwrap_alarm_destroy(grpcwrap_alarm* alarm) {
+  delete alarm;
+  return;
+}
+
+using grpc::internal::CompletionQueueTag;
+
+/** A helper class that implements `CompletionQueueTag`.
+ *
+ * TODO: Remove it. The `grpc::Alarm` submit itself as a tag to a
+ * `CompletionQueue`, so we need it to unify extraction of tags`.
+ */
+class GrpcwrapTag : public CompletionQueueTag {
```

Comment:
Is there any virtual methods in `CompletionQueueTag`? If so, maybe can't use `static_cast`.